### ISSUE TITLE
librdkafka potential memory leak?

### ISSUE THREAD

**Issue by yhsiung** [edenhill/librdkafka#3810]
Read the FAQ first: https://github.com/edenhill/librdkafka/wiki/FAQ

Do NOT create issues for questions, use the discussion forum: https://github.com/edenhill/librdkafka/discussions

Description
===========
My Kafka producer aborted at rd_malloc failure.  In the rd_kafka_msg_destroy() line 66, it call rd_free(rkm) only if the RD_KAFKA_MSG_F_FREE_RKM is set. However, my producer calls rd_kafka_produce() with RD_KAFKA_MSG_F_COPY only. Does that mean the "rkm" will not be freed after the dr_msg_cb() is called. which will eventually triggers rd_malloc() failure. My librdkafka is 1.2.2.

Do I need to set both RD_KAFKA_MSG_F_COPY and RD_KAFKA_MSG_F_FREE_RKM at the same time when calling rd_kafka_produce()?

How to reproduce
================
continue sending messages for a long time.

**IMPORTANT**: Always try to reproduce the issue on the latest released version (see https://github.com/edenhill/librdkafka/releases), if it can't be reproduced on the latest version the issue has been fixed.

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version (release number or git tag): 1.2.2
 - [ ] Apache Kafka version: `<REPLACE with e.g., 0.10.2.3>`
 - [x] librdkafka client configuration: message.timeout.ms=120000, batch.num.messages=5, queue.buffering.max.ms=100
 - [x] Operating system: Linux embedded system
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
FREE_RKM is an internal flag, only use the flags defined in rdkafka.h
1.2.2 is very old. Plesae upgrade to the latest release.

**Comment by yhsiung** [comment]
I checked the latest code. It has the same issue.

void rd_kafka_msg_destroy(rd_kafka_t *rk, rd_kafka_msg_t *rkm) {
        // FIXME
        if (rkm->rkm_flags & RD_KAFKA_MSG_F_ACCOUNT) {
                rd_dassert(rk || rkm->rkm_rkmessage.rkt);
                rd_kafka_curr_msgs_sub(rk ? rk : rkm->rkm_rkmessage.rkt->rkt_rk,
                                       1, rkm->rkm_len);
        }

        if (rkm->rkm_headers)
                rd_kafka_headers_destroy(rkm->rkm_headers);

        if (likely(rkm->rkm_rkmessage.rkt != NULL))
                rd_kafka_topic_destroy0(rkm->rkm_rkmessage.rkt);

        if (rkm->rkm_flags & RD_KAFKA_MSG_F_FREE && rkm->rkm_payload)
                rd_free(rkm->rkm_payload);

        if (rkm->rkm_flags & RD_KAFKA_MSG_F_FREE_RKM)
                rd_free(rkm);
}

**Comment by edenhill** [comment]
msgflags:
 * Use `RD_KAFKA_MSG_F_FREE` if you want librdkafka to free() the produce()d payload when delivery is done - librdkafka will take over ownership of the memory.
 * or use `RD_KAFKA_MSG_F_COPY` if you want librdkafka to make a copy of the payload and then free that when done - application has ownership of the memory and librdkafka does not use it.
 *  or specify no flags if you want to free the payload yourself - librdkafka takes over ownership of the memory until the delivery report callback where ownership is transferred back to the application.

**Comment by yhsiung** [comment]
I used the RD_KAFKA_MSG_F_COPY. My producer will free the payload after calling rd_kafka_produce(). However, my producer might get abort occasionally due to rd_malloc failure. 

rd_malloc: Assertion 'p' failed

What might cause it? Thank you.

**Comment by edenhill** [comment]
Upgrade to the latest version, we don't provide free support of old version.

**Comment by yhsiung** [comment]
The producer was aborted due to rd_malloc failure. I then relaunched the producer, but I got the "Local: Queue full" error immediately from librdkafka when the producer was trying to push new messages.  It seems that the old messages in the internal queue from the previous producer were not clear. Any ideas?

I have a question about rdkafka library. If I have two producers share (link to) the same librdkafka running in the same machine, do both producers share the same output queue in the librdkafka or separate output queue since they are running in different address space? In another word, is the internal queue in the shared memory? Thanks.
